### PR TITLE
Implement the Entry API for LinearMap

### PR DESCRIPTION
**Pull request by gereeter**

### CODE REVIEW

**Comment by FlashCat** [comment]
Thanks for the pull request, and welcome! The contain-rs team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

File: src/lib.rs [hunk 1]
```diff
@@ -218,6 +220,107 @@ impl<K:PartialEq+Eq,V> LinearMap<K,V> {
         }
         None
     }
+
+    /// Gets the given key's corresponding entry in the map for in-place manipulation.
+    pub fn entry(&mut self, key: K) -> Entry<K, V> {
+        for i in 0..self.storage.len() {
+            let found;
+            {
```
Comment (Gankro) [inline src/lib.rs]:
```
let found = {
    ...
    key == *k
};
```

Comment (Gankro) [inline src/lib.rs]: Actually can this all just be replaced with:

```
match self.storage.iter().position(|k| key == *k) {
  None => Vacant(..)
  Some(idx) => Occupied(..)
}
```

?

File: src/lib.rs [hunk 2]
```diff
@@ -469,6 +576,96 @@ mod test {
...
+            Vacant(view) => {
+                assert_eq!(*view.insert(1000), 1000);
+            }
+        }
+        assert_eq!(map.get(&10).unwrap(), &1000);
+        assert_eq!(map.len(), 6);
+    }
+
+    #[test]
+    fn test_entry_take_doesnt_corrupt() {
```
Comment (Gankro) [inline src/lib.rs]: I would drop this. It's a hashmap-specific regression test.

File: src/lib.rs [hunk 3]
```diff
@@ -300,8 +403,12 @@ impl<'a, K:'a, V:'a> ExactSizeIterator for Values <'a, K, V> { }
 #[cfg(test)]
 mod test {
     use super::LinearMap;
+    use super::Entry::{Occupied, Vacant};
 
     extern crate test;
+    extern crate rand;
+
+    use self::rand::{thread_rng, Rng};
```
Comment (Gankro) [inline src/lib.rs]: If you drop the hashmap test then you can drop `rand` too.

File: src/lib.rs (hunk 1 above)
Comment (gereeter) [inline src/lib.rs]: Err... yes, that does work. I just saw the `insert` implementation and forgot that `position` existed.

**Comment by gereeter** [comment]
I think I fixed everything.

**Comment by Gankro** [comment]
Sweet, thanks!

File: src/lib.rs (hunk 1 above)
Comment (Gankro) [inline src/lib.rs]: I will never forget the day I made a non-trivial improvement to *the* @gereeter's code.

Today is a good day.

**Comment by Gankro** [comment]
(waiting to hear back from travis)

**Comment by gereeter** [comment]
Hm... GitHub is still waiting for Travis to finish, but if you actually click on the link to Travis, it says that the tests passed a day ago.

**Comment by Gankro** [comment]
Weird.

**Pull request merged**